### PR TITLE
CB-1837 Update redbeams with working RDS instance launch

### PR DESCRIPTION
**Pull request by bhavanki**
Redbeams is now able to launch PostgreSQL RDS instances via
CloudFormation. The API endpoint for launching works as far as that, but
currently does not return successfully or register the created RDS
resources, effectively orphaning them. Near-term changes will rectify
these shortcomings, along with a good number of other missing features.

- The create endpoint in DatabaseServerV4Controller now calls on
  RedbeamsCreationService to start launching an RDS instance. It does
  not yet successfully return a response.
- AllocateDatabaseServerV4Request is expanded to include many necessary
  fields, including nested objects describing the network, database
  server, and security group information for the stack of resources to
  be created. Only AWS stacks are currently supported.
- AllocateDatabaseServerV4Request can now be converted into a DBStack
  entity. This conversion occurs early in the API call to create a new
  database server.
- The following fields are added to DBStack: region, availabilityZone,
  cloudPlatform, platformVariant, template, and ownerCrn. A
  corresponding migration is included.
- The launchDatabase method of RedbeamsCreationService is now
  implemented. It loads the provider-specific template for database
  server creation and saves it on the DBStack, saves the DBStack, and
  then sends the event necessary to start the flow for database server
  creation. It does not yet return a correct outcome response.
- DBStackService and DBStackRepository are introduced to enable working
  with DBStack entities in the redbeams database.
- The Credential entity has been changed into an ordinary data transfer
  object (DTO), and its table is removed from the redbeams database.
- The new EnvironmentConfig configuration object defines parameters
  necessary for redbeams to call the environment service to get
  environment information and credentials, using EnvironmentService and
  CredentialsService, respectively. The environment-api module is now a
  dependency for redbeams. ServiceEndpointConfig makes the environment
  service URL available.
- The existing SecretAspectService, which is responsible for managing
  vault entries for secret values such as usernames and passwords, is
  insufficient for the DBStack entity, because the database server root
  username and password reside in a nested DatabaseServer entity which
  lacks its own repository class. The new RecursiveSecretAspectService
  remedies this shortcoming, by altering the field scan of entities to
  also recurse into nested ones that are annotated with @SecretValue.
- RedbeamsFlowManager now passes the user CRN stored in the current
  thread to events that it sends as notifications. This provides the
  user CRN to code run in flows that need to call to other services that
  require the user CRN, such as the environment service.
- CrnService can now return the current user ID.
- RedbeamsContext now includes the DatabaseStack object used as the
  basis for provider-specific work.
- The createFlowContext method of AbstractRedbeamsProvisionAction is now
  implemented. It is able to build a RedbeamsContext that is carried on
  events sent to flow handlers. RedbeamsProvisionActions completes the
  connection for database server allocations by constructing a newly
  expanded and sufficient AllocateDatabaseServerRequest event.
- Flow events AllocateDatabaseServerSuccess and
  AllocateDatabaseServerFailed are now subclasses of
  AllocateDatabaseServerResponse. AllocateDatabaseServerFailed now
  includes error details, and AllocateDatabaseServerSuccess now includes
  a list of cloud resource status results.
- AllocateDatabaseServerHandler is fully implemented. It asks a
  provider-specific resource connector to launch a new database server,
  and polls the launch process until completion, sending events for
  success and failure.
- AwsRdsLaunchService no longer expects a VPC ID to be part of the
  network information for a new RDS instance.
- The outputs of the CloudFormation template for new database stacks are
  fixed.

These changes outside redbeams are included.

- CloudPlatformVariant gains a new constructor that accepts strings.
- The Location class in cloud-api now has equals and hashCode methods.
- The ResourceBuilders component in cloud-template no longer requires at
  least one NetworkResourceBuilder, ComputeResourceBuilder, and
  GroupResourceBuilder.

### CODE REVIEW

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsLaunchService.java [hunk 1]
```diff
@@ -81,10 +81,11 @@
             cfRetryClient.describeStacks(new DescribeStacksRequest().withStackName(cFStackName));
             LOGGER.debug("Stack already exists: {}", cFStackName);
         } catch (AmazonServiceException ignored) {
-            boolean existingVPC = awsNetworkView.isExistingVPC();
+            // boolean existingVPC = awsNetworkView.isExistingVPC();
```
Comment (jadair-cloudera) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsLaunchService.java]: What is the plan here?

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/requests/AllocateDatabaseServerV4Request.java [hunk 2]
```diff
@@ -1,10 +1,131 @@
...
+import javax.validation.constraints.Size;
 
 @ApiModel
 @JsonIgnoreProperties(ignoreUnknown = true)
-public class AllocateDatabaseServerV4Request {
+public class AllocateDatabaseServerV4Request extends ProviderParametersBase {
+
+    // FIXME Define ModelDescriptions.DBStackModelDescription
+
+    @Size(max = 40, min = 5, message = "The length of the name has to be in range of 5 to 40")
```
Comment (jadair-cloudera) [inline redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/requests/AllocateDatabaseServerV4Request.java]: NIT: grammar

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/requests/AllocateDatabaseServerV4Request.java [hunk 3]
```diff
@@ -1,10 +1,131 @@
...
+    @NotNull
+    @ApiModelProperty(value = /* DBStackModelDescription.STACK_NAME */ "name of the stack", required = true)
+    private String name;
+
+    @NotNull
+    @ApiModelProperty(value = /* StackModelDescription.ENVIRONMENT_CRN */ "ID of the environment", required = true)
+    private String environmentId;
+
+    @NotNull
+    @ApiModelProperty(value = "Region for database stack")
```
Comment (jadair-cloudera) [inline redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/requests/AllocateDatabaseServerV4Request.java]: Why no comment like the others?

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsLaunchService.java (hunk 1 above)
Comment (bhavanki) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsLaunchService.java]: Not sure ... I'll think about it. RDS doesn't require you to state a VPC, but maybe it should be present for some other reason.

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/requests/AllocateDatabaseServerV4Request.java (hunk 3 above)
Comment (bhavanki) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/requests/AllocateDatabaseServerV4Request.java]: Laziness

**Comment by bhavanki** [comment]
The previous unit test run timed out, but a retry succeeded.

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsLaunchService.java (hunk 1 above)
Comment (bhavanki) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsLaunchService.java]: I will address this in a follow-up change.

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/requests/AllocateDatabaseServerV4Request.java (hunk 2 above)
Comment (bhavanki) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/requests/AllocateDatabaseServerV4Request.java]: I will address this in a follow-up change.

**Pull request merged**